// mtl/ui/SelectorPlugin.js
var	Selection = require('mtl/ui/Selection'),
	Point = require('mtl/ui/Point'),
	Matcher = require('mtl/ui/Matcher'),
	Highlighter = require('mtl/ui/Highlighter');


module.exports = SelectorPlugin;


var VK_ALT = 18,
	VK_ESC = 27;

var html = document.documentElement;


function SelectorPlugin() {
	this.onGlobalMouseDown_ = this.onGlobalMouseDown_.bind(this);
	this.onGlobalMouseUp_ = this.onGlobalMouseUp_.bind(this);
	this.onGlobalKeyUp_ = this.onGlobalKeyUp_.bind(this);
}

var p = SelectorPlugin.prototype;

p.document_ = null;
p.html_ = null;
p.selection_ = null;
p.overlay_ = null;

p.attach = function(document) {
	this.document_ = document;
	this.html_ = document.documentElement;

	this.document_.addEventListener('mousedown', this.onGlobalMouseDown_, true);
};

p.detach = function() {
	this.document_.removeEventListener('mousedown', this.onGlobalMouseDown_, true);

	this.document_ = null;
	this.html_ = null;
};

p.onGlobalMouseDown_ = function (event) {
	if (!event.altKey) {
		return;
	}

	event.preventDefault();
	event.stopPropagation();

	this.startSelection_(Point.fromEventClientXY(event));
};

p.onGlobalMouseUp_ = function (event) {
	event.preventDefault();
	event.stopPropagation();

	this.doneSelection_();
};

p.onGlobalKeyUp_ = function (event) {
	if (event.keyCode === VK_ALT) {
		this.cancelSelection_();
	}
};

p.createOverlay_ = function () {
	var doc = this.document_;

	var overlay = doc.createElement('div');
	var style = overlay.style;

	style.position = 'fixed';
	style.display = 'block';
	style.top = 0;
	style.right = 0;
	style.bottom = 0;
	style.left = 0;
	style.cursor = 'crosshair';
	style.zIndex = 16777271;

	return overlay;
};

p.getOverlay_ = function () {
	if (!this.overlay_) {
		this.overlay_ = this.createOverlay_();
	}

	return this.overlay_;
};

p.showOverlay_ = function () {
	var overlay = this.getOverlay_();
	this.html_.appendChild(overlay);

	return overlay;
};

p.hideOverlay_ = function () {
	var overlay = this.getOverlay_();

	if (overlay.parentNode) {
		overlay.parentNode.removeChild(overlay);
	}

	return overlay;
};

p.startSelection_ = function (startPoint) {
	var html = this.html_;

	html.addEventListener('mouseup', this.onGlobalMouseUp_, true);
	html.addEventListener('keyup', this.onGlobalKeyUp_, true);

	var overlay = this.showOverlay_();

	this.selection_ = new Selection(overlay, new Matcher(overlay), new Highlighter(overlay));
	this.selection_.start(startPoint);
};

p.cancelSelection_ = function () {
	this.endSelection_();
};

p.doneSelection_ = function () {
	var selectedLinks = this.selection_.getSelectedLinks();
	this.endSelection_();

	if (this.onselected) {
		this.onselected(selectedLinks);
	}
};

p.endSelection_ = function () {
	var html = this.html_;

	html.removeEventListener('mouseup', this.onGlobalMouseUp_, true);
	html.removeEventListener('keyup', this.onGlobalKeyUp_, true);

	this.selection_.end();
	this.hideOverlay_();

	this.selection_.release();
};
